src/libraries/Common/src/System/Security/Cryptography/RSASecurityTransforms.macOS.cs

```diff
@@ -0,0 +1,79 @@
+using System.Security.Cryptography.Asn1;
+using Internal.Cryptography;
+
+namespace System.Security.Cryptography
+{
+    internal static partial class RSAImplementation
+    {
+        public sealed partial class RSASecurityTransforms : RSA
+        {
+            private static RSAParameters ExportParametersFromLegacyKey(SecKeyPair keys, bool includePrivateParameters)
+            {
+                // Apple requires all private keys to be exported encrypted, but since we're trying to export
+                // as parsed structures we will need to decrypt it for the user.
+                const string ExportPassword = "DotnetExportPassphrase";
+
+                byte[] keyBlob = Interop.AppleCrypto.SecKeyExport(
+                    includePrivateParameters ? keys.PrivateKey : keys.PublicKey,
+                    exportPrivate: includePrivateParameters,
+                    password: ExportPassword);
+
+                try
+                {
+                    if (!includePrivateParameters)
+                    {
+                        // When exporting a key handle opened from a certificate, it seems to
+                        // export as a PKCS#1 blob instead of an X509 SubjectPublicKeyInfo blob.
+                        // So, check for that.
+                        // NOTE: It doesn't affect macOS Mojave when SecCertificateCopyKey API
+                        // is used.
+                        RSAParameters key;
+
+                        AsnReader reader = new AsnReader(keyBlob, AsnEncodingRules.BER);
+                        AsnReader sequenceReader = reader.ReadSequence();
+
+                        if (sequenceReader.PeekTag().Equals(Asn1Tag.Integer))
+                        {
+                            AlgorithmIdentifierAsn ignored = default;
+                            RSAKeyFormatHelper.ReadRsaPublicKey(keyBlob, ignored, out key);
+                        }
+                        else
+                        {
+                            RSAKeyFormatHelper.ReadSubjectPublicKeyInfo(
+                                keyBlob,
+                                out int localRead,
+                                out key);
+                            Debug.Assert(localRead == keyBlob.Length);
+                        }
+                        return key;
+                    }
+                    else
+                    {
+                        RSAKeyFormatHelper.ReadEncryptedPkcs8(
+                            keyBlob,
+                            ExportPassword,
+                            out int localRead,
+                            out RSAParameters key);
+                        return key;
+                    }
+                }
+                finally
+                {
+                    CryptographicOperations.ZeroMemory(keyBlob);
+                }
+            }
+
+            private static bool HasWorkingPKCS1Padding { get; } = OperatingSystem.IsMacOSVersionAtLeast(10, 15);
```

Comment:
What's the behavior on 10.13/10.14?  Can we get away with, e.g., using the native decrypt and when it fails then check to see if it's a validly padded zero byte message?

Alternatively, maybe we just don't use private data keys on macOS before 10.15.

Reply:
It reports `errSecParam` error which is the same for any invalid padding and the valid padding of zero-sized message. I was doing the additional check but I felt I'd end up in some timing attack trap (which I ended up in anyway).

Using the old-style RSA keys is definitely an option. The legacy export code is there anyway (for S.S.C.X509Certificates) and import is nearly identical since both APIs use PKCS#1 private key blobs. I think I prefer that to rolling out my own padding code and trying to get constant time (which the old macOS code doesn't do anyway; but new one does).

Reply:
Note that there is a very small chance that old iOS systems had the same bug. I have no way to test it. In that case there's no fallback to the legacy CSSM keys.